### PR TITLE
AST refactoring

### LINKED ISSUE #330: Discussion for AST simplification
The summer has seen a thourough code reorganisation, with the unification of shared code and types for different passes under the `shared_ast` sub-library.

However, the AST structures themselves have been left unchanged, if defined in a different way (one thing at a time!) . For example, while the type for tuples and structures is now separated, but the underlying expressions still fall into the same `EStruct` case with an option for the structure name.

This is to open a thread for discussion on the changes that we want or not before actually implementing the changes. I'll write different aspects in different comments.

### PR DESCRIPTION
**Pull request by AltGr**
Many changes got bundled in here and would be too tedious to separate.

Closes #330

See changes in `shared_ast/definitions.ml` to check the main point.

- the biggest change is a modification of the struct and enum types in
 expressions: they are now stored as `Map`s throughout passes, and no
longer
 converted to indexed lists after scopelang. Their accessors are also
changed,
 and tuples only exist in Lcalc (they're used for closure conversion).

  This implied adding some more information in the contexts, to keep the
mapping
 between struct fields and scope output variables. It should also be much
more
 robust (no longer relying on assumptions upon different orderings).

- another very pervasive change is more cosmetic: the rewrite of the main
AST to
 use inline records, labelling individual subfields.

- moved the checks for correct definitions and accesses of structures from
 `Scope_to_dcalc` to `Typing`

- defining some new shallow iterators in module `Shared_ast.Expr`, and
 factorising a few same-pass rewriting functions accordingly (closure
 conversion, optimisations, etc.)

- some smaller style improvements (ensuring we use the proper compare/equal
 functions instead of `=` in a few `when` closes, for example)

### CODE REVIEW

File: compiler/scalc/to_python.ml [hunk 1]
```diff
@@ -461,40 +463,41 @@ let format_ctx
       struct_name
       (Format.pp_print_list
          ~pp_sep:(fun fmt () -> Format.fprintf fmt ", ")
-         (fun _fmt (struct_field, struct_field_type) ->
+         (fun fmt (struct_field, struct_field_type) ->
```
Comment (denismerigoux) [inline compiler/scalc/to_python.ml:466]: wouahou good catch

File: compiler/scopelang/scope_to_dcalc.ml [hunk 2]
```diff
@@ -531,25 +500,31 @@ let translate_rule
...
+                 subscope_sig.scope_sig_in_fields)
+          in
+          StructFieldMap.add field e acc)
+        StructFieldMap.empty all_subscope_input_vars
     in
     let subscope_struct_arg =
-      (* FIXME: this is very fragile: we assume that the ordering of the scope
-         variables is the same as the ordering of the struct fields. *)
-      Expr.etuple subscope_args (Some called_scope_input_struct)
+      Expr.estruct called_scope_input_struct subscope_args
```
Comment (denismerigoux) [inline compiler/scopelang/scope_to_dcalc.ml:527]: YEAH

File: compiler/verification/z3backend.real.ml [hunk 3]
```diff
@@ -313,11 +314,17 @@ and find_or_create_enum (ctx : context) (enum : EnumName.t) :
...
+        (fun ctr ty (ctx, ctrs) ->
+          let ctx, ctr = create_constructor ctr ty ctx in
+          ctx, ctr :: ctrs)
+        ctrs (ctx, [])
+    in
     let z3_enum =
       Datatype.mk_sort_s ctx.ctx_z3
         (Marked.unmark (EnumName.get_info enum))
-        z3_ctrs
+        (List.rev z3_ctrs)
```
Comment (denismerigoux) [inline compiler/verification/z3backend.real.ml:327]: `fold_left` vs `fold_right` semantics for `EnumConstructorMap.fold` ?

File: tests/test_array/good/aggregation_2.catala_en [hunk 4]
```diff
@@ -29,13 +29,13 @@ scope B:
 $ catala Interpret -s A
 [RESULT] Computation successful! Results:
 [RESULT] x =
-           [S {"id"= 0; "income"= $0.00}; S {"id"= 1; "income"= $9.00};
-             S {"id"= 2; "income"= $5.20}]
+           [ S { "id"= 0; "income"= $0.00 };  S { "id"= 1; "income"= $9.00 };
```
Comment (denismerigoux) [inline tests/test_array/good/aggregation_2.catala_en:32]: Why the extra space here ?

Comment (denismerigoux) [reply tests/test_array/good/aggregation_2.catala_en:32]: especially before the struct name

**Review by denismerigoux — APPROVED** [review]
Thanks again for this massive PR!

**Comment by denismerigoux** [comment]
Before merging, can you `make format` the generated OCaml and Python files in `french_law` ?

File: compiler/verification/z3backend.real.ml (hunk 3 above)
Comment (AltGr) [reply compiler/verification/z3backend.real.ml:327]: Yeah this one actually took me a little extra time to debug :/
`Set`/`Map` always fold from the left, and don't even provide reverse fold functions (although they wouldn't be any more expensive...) ; so the fold actually returns a reversed list.

Best solution would probably be to have a `Map.fold_map` function that already preserves order...

File: tests/test_array/good/aggregation_2.catala_en (hunk 4 above)
Comment (AltGr) [reply tests/test_array/good/aggregation_2.catala_en:32]: Good remark, that's suspicious indeed, I'll see how the printer got changed.

**Comment by AltGr** [comment]
> Before merging, can you make format the generated OCaml and Python files in french_law ?

hm I am not sure what you mean, we disabled reformatting of generated files a while ago :grimacing:

**Pull request merged**